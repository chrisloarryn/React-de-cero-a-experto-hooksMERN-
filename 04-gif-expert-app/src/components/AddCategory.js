import React, { useState } from 'react'
import PropTypes from 'prop-types'

export const AddCategory = ({ setCategories }) => {
  const [inputValue, setInputValue] = useState('Hello World')

  const handleInputChange = e => {
    e.preventDefault()
    setInputValue(e.target.value)
    // console.log('handleInputChange invoked')
  }

  const handleSubmit = e => {
    e.preventDefault()
    // console.log('handleSubmit', inputValue)
    setCategories(cats => [{ name: inputValue }, ...cats])
    setInputValue('')
  }

  return (
    <form onSubmit={handleSubmit}>
      <p>{inputValue}</p>
      <input type='text' value={inputValue} onChange={handleInputChange} />
    </form>
  )
}

AddCategory.propTypes = {
  setCategories: PropTypes.func.isRequired
}
