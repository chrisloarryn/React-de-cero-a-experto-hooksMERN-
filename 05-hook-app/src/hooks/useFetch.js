import { useState, useEffect, useRef } from 'react'

export const useFetch = url => {
  const isMounted = useRef(true)
  const [state, setState] = useState({
    data: null,
    loading: true,
    error: null
  })
  useEffect(() => {
    return () => {
      isMounted.current = false
    }
  }, [])
  useEffect(() => {
    setState({ data: null, loading: true, error: null })
    fetch(url)
      .then(resp => resp.json())
      .then(data => {
        setTimeout(() => {
          if (isMounted.current)
            setState({
              loading: false,
              error: null,
              data
            })
          else console.log('setState() did not called')
        }, 1000)
      })
      .catch(() => {
        setState({
          data: null,
          loading: false,
          error: 'data has not been charged'
        })
      })
  }, [url])

  return state
}
