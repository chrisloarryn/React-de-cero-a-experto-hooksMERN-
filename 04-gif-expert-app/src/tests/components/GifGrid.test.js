import React from 'react'
import '@testing-library/jest-dom'
import { shallow } from 'enzyme'

import { GifGrid } from './../../components/GifGrid'
import { useFetchGifs } from '../../hooks/useFetchGifs'
jest.mock('../../hooks/useFetchGifs')

describe(`Test into <GifGrid /> component`, () => {
  const category = 'One Punch'

//   beforeEach(() => {
//     wrapper = shallow(<GifGrid category={category} />)
//   })

  test('test into the component', () => {
    useFetchGifs.mockReturnValue({
      data: [],
      loading: true
    })
    const wrapper = shallow(<GifGrid category={category} />)
    expect(wrapper).toMatchSnapshot()
  })

  test('should show items when useFetchGifs load them', () => {
    const gifs = [
      {
        id: 'ABC',
        url: 'https://localhost/any/thing.jpg',
        title: 'Anything'
      },
      {
        id: '123',
        url: 'https://localhost/any/thing.jpg',
        title: 'Anything'
      }
    ]
    useFetchGifs.mockReturnValue({
      data: gifs,
      loading: false
    })
    const wrapper = shallow(<GifGrid category={category} />)
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find('p').exists()).toBe(false)
    expect(wrapper.find('GifGridItem').length).toBe(gifs.length)
  })
})
