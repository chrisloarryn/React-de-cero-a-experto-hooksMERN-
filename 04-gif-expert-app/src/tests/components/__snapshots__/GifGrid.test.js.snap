// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test into <GifGrid /> component should show items when useFetchGifs load them 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__bounceIn animate__delay-1s"
  >
    One Punch
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="ABC"
      key="ABC"
      title="Anything"
      url="https://localhost/any/thing.jpg"
    />
    <GifGridItem
      id="123"
      key="123"
      title="Anything"
      url="https://localhost/any/thing.jpg"
    />
  </div>
</Fragment>
`;

exports[`Test into <GifGrid /> component test into the component 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__bounceIn animate__delay-1s"
  >
    One Punch
  </h3>
  <p
    className="animate__animated animate__flash"
  >
    Loading...
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;
